File: test/python_parity_generate.jl
# Live Python-parity check for generate / generate_inv with a NON-IDENTITY indices.
# Opt-in (needs the fixture from python_parity_generate.py); not part of the default suite.
#
#   python test/python_parity_generate.py test/reference/generate_parity.npz       # graphgp env
#   julia --project=bench test/python_parity_generate.jl \
#       test/reference/generate_parity.npz                                          # needs NPZ
using GraphGP, NPZ, Printf

path = length(ARGS) >= 1 ? ARGS[1] : joinpath(@__DIR__, "reference", "generate_parity.npz")
d = npzread(path)

coords = permutedims(UInt32.(d["coords"]))            # (N,d) -> (d,N), tree order
neighbors = permutedims(Int.(d["neighbors"])) .+ 1    # (M,k) 0-based -> (k,M) 1-based
offsets = Int.(d["offsets"])
n0 = Int(d["n0"]); scale = Float64(d["scale"])
indices = Int.(d["indices"]) .+ 1                     # 0-based tree→orig -> 1-based
bins = Float64.(d["cov_bins64"]); vals = Float64.(d["cov_vals64"])
prob = GraphGPProblem(coords, neighbors, offsets, n0, scale, bins, vals, indices)

xi = Float64.(d["xi"]); values = Float64.(d["values"])
field_py = Float64.(d["generate_field"]); xi_py = Float64.(d["generate_inv_xi"])

field_jl = generate(prob, xi)                         # ORIGINAL order in/out
xi_jl = generate_inv(prob, values)                   # ORIGINAL order in/out

relerr(a, b) = sqrt(sum(abs2, a .- b)) / max(sqrt(sum(abs2, b)), 1e-30)
e_gen = relerr(field_jl, field_py)
e_inv = relerr(xi_jl, xi_py)
e_round = relerr(generate_inv(prob, field_jl), xi)

@printf("N=%d  non-identity indices=%s\n", length(xi), indices != collect(1:length(xi)))
@printf("  generate     vs Python: norm-relerr = %.2e\n", e_gen)
@printf("  generate_inv vs Python: norm-relerr = %.2e\n", e_inv)
@printf("  roundtrip generate_inv∘generate (original order): norm-relerr = %.2e\n", e_round)
ok = e_gen < 1e-12 && e_inv < 1e-12 && e_round < 1e-10
println(ok ? "PARITY PASS" : "PARITY FAIL")
exit(ok ? 0 : 1)

File: test/python_parity_generate.py
"""Dump a Python `graphgp` generate / generate_inv reference with a NON-IDENTITY `indices`, so
the Julia drop-in can be checked element-wise in ORIGINAL point order.

    python python_parity_generate.py [out.npz]

`graphgp.generate(graph, cov, xi)` and `generate_inv(graph, cov, values)` are both original-order
in / original-order out. We dump the integer-lattice graph (tree order) + `graph.indices`, a
random original-order `xi` and its field, and a random original-order `values` and its xi.
Run with an env that has `graphgp` (e.g. ~/.venv-gpu). Checked by python_parity_generate.jl.
"""
import sys

import jax
jax.config.update("jax_enable_x64", True)
import jax.numpy as jnp
import jax.random as jr
import numpy as np
import graphgp as gp

out = sys.argv[1] if len(sys.argv) > 1 else "reference/generate_parity.npz"
N, D, n0, k = 2000, 3, 200, 8
BITS = 21
LMAX = (1 << BITS) - 1

key = jr.key(20240607)
kp, kx, kv = jr.split(key, 3)
points = np.asarray(jr.normal(kp, (N, D), dtype=jnp.float64))
origin = points.min(axis=0)
scale = float((points.max(axis=0) - origin).max()) / LMAX
coords0 = np.clip(np.rint((points - origin) / scale), 0, LMAX).astype(np.uint32)
points_q = origin + scale * coords0.astype(np.float64)

graph = gp.build_graph(jnp.asarray(points_q, dtype=jnp.float64), n0=n0, k=k)
cov = gp.extras.rbf_kernel(variance=1.0, scale=0.3, r_min=1e-4, r_max=1e1, n_bins=400, jitter=1e-3)
cov = (jnp.asarray(cov[0], jnp.float64), jnp.asarray(cov[1], jnp.float64))

# Integer coords in tree order, and the (0-based) tree→original permutation.
coords = np.rint((np.asarray(graph.points, dtype=np.float64) - origin) / scale).astype(np.uint32)
indices = np.asarray(graph.indices, dtype=np.int64)            # indices[tree_pos] = original_pos

xi = np.asarray(jr.normal(kx, (N,), dtype=jnp.float64))        # ORIGINAL order
field = np.asarray(gp.generate(graph, cov, jnp.asarray(xi)))   # ORIGINAL order
values = np.asarray(jr.normal(kv, (N,), dtype=jnp.float64))    # ORIGINAL order
xi_from_values = np.asarray(gp.generate_inv(graph, cov, jnp.asarray(values)))  # ORIGINAL order

assert not np.array_equal(indices, np.arange(N)), "indices is identity — test would be vacuous"
np.savez(
    out,
    coords=coords, neighbors=np.asarray(graph.neighbors, dtype=np.int64),
    offsets=np.asarray(graph.offsets, dtype=np.int64), n0=np.int64(n0),
    scale=np.float64(scale), indices=indices,
    cov_bins64=np.asarray(cov[0]), cov_vals64=np.asarray(cov[1]),
    xi=xi, generate_field=field, values=values, generate_inv_xi=xi_from_values,
)
print(f"dumped generate parity (N={N}, non-identity indices) -> {out}")
